File: src/main/java/com/now/nowbot/service/MessageServiceImpl/CustomService.java
package com.now.nowbot.service.MessageServiceImpl;

import com.now.nowbot.config.FileConfig;
import com.now.nowbot.dao.BindDao;
import com.now.nowbot.mapper.UserProfileMapper;
import com.now.nowbot.model.BinUser;
import com.now.nowbot.qq.event.MessageEvent;
import com.now.nowbot.qq.message.ImageMessage;
import com.now.nowbot.qq.message.ReplyMessage;
import com.now.nowbot.service.ImageService;
import com.now.nowbot.service.MessageService;
import com.now.nowbot.throwable.ServiceException.BindException;
import com.now.nowbot.throwable.ServiceException.CustomException;
import com.now.nowbot.util.ASyncMessageUtil;
import com.now.nowbot.util.DataUtil;
import com.now.nowbot.util.Instructions;
import com.now.nowbot.util.QQMsgUtil;
import jakarta.annotation.Resource;
import org.apache.logging.log4j.util.Strings;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.lang.NonNull;
import org.springframework.lang.Nullable;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.nio.file.Files;
import java.nio.file.NoSuchFileException;
import java.nio.file.Path;
import java.util.Objects;

@Service("CUSTOM")
public class CustomService implements MessageService<CustomService.CustomParam> {
    private static final Logger log = LoggerFactory.getLogger(CustomService.class);
    private static Path FILE_DIV_PATH;
    @Resource
    RestTemplate restTemplate;
    @Resource
    BindDao bindDao;
    @Resource
    ImageService imageService;
    @Resource
    UserProfileMapper userProfileMapper;

    @Autowired
    public CustomService(FileConfig fileConfig) {
        FILE_DIV_PATH = Path.of(fileConfig.getBgdir(), "user-profile");
    }

    public record CustomParam(@NonNull long uid, @NonNull Type type, @Nullable String url) {
        //url == null 是删除图片
    }

    @Override
    public boolean isHandle(MessageEvent event, String messageText, DataValue<CustomParam> data) throws Throwable {
        var from = event.getSubject();

        var matcher2 = Instructions.DEPRECATED_SET.matcher(messageText);
        if (matcher2.find() && Strings.isNotBlank(matcher2.group("set"))) {
            throw new CustomException(CustomException.Type.CUSTOM_Instruction_Deprecated);
        }

        var matcher = Instructions.CUSTOM.matcher(messageText);
        if (! matcher.find()) {
            return false;
        }

        BinUser u;
        try {
            u = bindDao.getUserFromQQ(event.getSender().getId());
        } catch (BindException e) {
            throw new CustomException(CustomException.Type.CUSTOM_Me_TokenExpired);
        }

        var firstMessage = event.getMessage().getMessageList().getFirst();

        var save = matcher.group("save");
        String imgPath;

        if (Objects.nonNull(save) && ! (save.equals("s") || save.equals("save") || save.equals("a") || save.equals("add"))) {
            // 需要删除
            imgPath = null;
        } else if (! (firstMessage instanceof ReplyMessage reply)) {
            // 不是回复。发送引导
            try {
                var md = DataUtil.getMarkdownFile("Help/custom.md");
                var image = imageService.getPanelA6(md, "help");
                from.sendImage(image);
                return false;
            } catch (Exception e) {
                throw new CustomException(CustomException.Type.CUSTOM_Instructions);
            }
        } else {
            // 正常
            if (Objects.isNull(event.getBot())) {
                throw new CustomException(CustomException.Type.CUSTOM_Receive_NoBot);
            }

            var msg = event.getBot().getMessage(reply.getId());
            ImageMessage img;

            if (Objects.isNull(msg) || Objects.isNull(img = QQMsgUtil.getType(msg, ImageMessage.class))) {
                //消息为空，并且不是回复的图片。询问是否删除
                var receipt = from.sendMessage(CustomException.Type.CUSTOM_Question_Clear.message);

                var lock = ASyncMessageUtil.getLock(event, 30 * 1000);
                event = lock.get();

                if (Objects.isNull(event) || ! event.getRawMessage().toUpperCase().contains("OK")) {
                    //不删除。失败撤回
                    from.recall(receipt);
                    return false;
                } else {
                    //确定删除
                    from.recall(receipt);
                    imgPath = null;
                }
            } else {
                //成功
                imgPath = img.getPath();
            }
        }

        switch (matcher.group("type")) {
            case "c", "card" -> data.setValue(
                    new CustomParam(u.getOsuID(), Type.card, imgPath)
            );
            case null, default -> data.setValue(
                    new CustomParam(u.getOsuID(), Type.banner, imgPath)
            );
        }

        return true;
    }

    @Override
    public void HandleMessage(MessageEvent event, CustomParam param) throws Throwable {
        var fileName = STR."\{param.uid}-\{param.type}.png";
        Path path = FILE_DIV_PATH.resolve(fileName);

        byte[] data = new byte[0];

        if (Objects.nonNull(param.url)) {
            data = restTemplate.getForObject(param.url, byte[].class);
            if (Objects.isNull(data)) {
                throw new CustomException(CustomException.Type.CUSTOM_Receive_PictureFetchFailed);
            }
        }

        var profile = userProfileMapper.getProfileById(param.uid());

        if (Objects.nonNull(param.url)) {
            // 保存
            try {
                Files.write(path, data);
            } catch (Exception e) {
                log.error("自定义：文件添加失败", e);
                throw new CustomException(CustomException.Type.CUSTOM_Set_Failed, param.type);
            }
            switch (param.type) {
                case card -> profile.setCard(path.toAbsolutePath().toString());
                case banner -> profile.setBanner(path.toAbsolutePath().toString());
            }
            userProfileMapper.saveAndFlush(profile);
            throw new CustomException(CustomException.Type.CUSTOM_Set_Success, param.type);
        } else {
            // 删除
            try {
                Files.delete(path);
            } catch (NoSuchFileException e) {
                throw new CustomException(CustomException.Type.CUSTOM_Clear_NoSuchFile, param.type);
            } catch (Exception e) {
                log.error("自定义：文件删除失败", e);
                throw new CustomException(CustomException.Type.CUSTOM_Clear_Failed, param.type);
            }
            switch (param.type) {
                case card -> profile.setCard(null);
                case banner -> profile.setBanner(null);
            }
            userProfileMapper.saveAndFlush(profile);
            throw new CustomException(CustomException.Type.CUSTOM_Clear_Success, param.type);
        }
    }

    enum Type {
        banner,
        card,
        header,
        info,
        score,
        ppm,
    }
}

File: src/main/java/com/now/nowbot/service/messageServiceImpl/CustomService.kt
package com.now.nowbot.service.messageServiceImpl

import com.now.nowbot.config.FileConfig
import com.now.nowbot.dao.BindDao
import com.now.nowbot.entity.ServiceCallStatistic
import com.now.nowbot.entity.UserProfileItem
import com.now.nowbot.entity.UserProfileKey
import com.now.nowbot.entity.UserProfileLite
import com.now.nowbot.mapper.UserProfileItemRepository
import com.now.nowbot.mapper.UserProfileRepository
import com.now.nowbot.model.UserProfile
import com.now.nowbot.model.UserProfile.UserProfileType.*
import com.now.nowbot.qq.event.MessageEvent
import com.now.nowbot.service.MessageService
import com.now.nowbot.service.MessageService.DataValue
import com.now.nowbot.service.messageServiceImpl.CustomService.CustomParam
import com.now.nowbot.service.messageServiceImpl.CustomService.CustomOperate.*
import com.now.nowbot.throwable.TipsException
import com.now.nowbot.throwable.botRuntimeException.IllegalStateException
import com.now.nowbot.util.ASyncMessageUtil
import com.now.nowbot.util.Instruction
import okio.IOException
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.web.reactive.function.client.WebClient
import java.nio.file.Files
import java.nio.file.NoSuchFileException
import java.nio.file.Path
import java.util.regex.Matcher

@Service("CUSTOM")
class CustomService(
    private val webClient: WebClient,
    private val bindDao: BindDao,
    private val userProfileRepository: UserProfileRepository,
        // todo: 等待数据迁移后使用
    private val userProfileItemRepository: UserProfileItemRepository,
    fileConfig: FileConfig,
) : MessageService<CustomParam> {

    init {
        FILE_DIV_PATH = Path.of(fileConfig.bgdir, "user-profile")
    }

    // url == null 是删除图片
    data class CustomParam(
        val uid: Long,
        val type: UserProfile.UserProfileType,
        val url: String?,
    )

    override fun isHandle(event: MessageEvent, messageText: String, data: DataValue<CustomParam>): Boolean {

        val matcher = Instruction.CUSTOM.matcher(messageText)
        if (!matcher.find()) {
            return false
        }

        data.value = getParam(event, matcher)

        return true
    }

    private fun getParam(event: MessageEvent, matcher: Matcher): CustomParam {
        val operate = CustomOperate.getOperate(matcher.group("operate"))
        val type = UserProfile.UserProfileType.getType(matcher.group("type"))

        val id = bindDao.getBindFromQQ(event.sender.id).userID

        return when (operate) {
            ADD -> if (event.hasImage() && !event.image?.path.isNullOrEmpty()) {
                val url = event.image!!.path!!

                CustomParam(id, type, url)
            } else {
                waitImage(id, type, event)
            }

            DELETE -> confirmDelete(id, type, event)
            UNKNOWN -> waitImage(id, type, event)
        }
    }

    private fun confirmDelete(id: Long, type: UserProfile.UserProfileType, event: MessageEvent): CustomParam {
        val receipt =
            event.reply("""
                您想要清除你的自定义 ${type.name} 吗？回复 OK 确认。
                如果并不想，请无视。
                """.trimIndent())

        val lock = ASyncMessageUtil.getLock(event, 60 * 1000)
        val ev = lock.get()

        if (ev != null && ev.rawMessage.contains("OK", ignoreCase = true)) {

            receipt.recall()
            return CustomParam(id, type, null)
        } else {
            receipt.recall()
            throw TipsException("清除过程已中止。")
        }
    }

    private fun waitImage(id: Long, type: UserProfile.UserProfileType, event: MessageEvent): CustomParam {
        // 消息为空，并且不是回复的图片。询问是否删除
        val receipt =
            event.reply("请发送您需要上传的图片 (${type.name})。")

        val lock = ASyncMessageUtil.getLock(event, 60 * 1000)
        val ev = lock.get()

        if (ev != null && ev.hasImage() && !ev.image?.path.isNullOrEmpty()) {
            val url = ev.image!!.path!!


            receipt.recall()
            return CustomParam(id, type, url)
        } else {

            receipt.recall()
            throw TipsException("没有获取到您发送的图片。请重试。")
        }
    }

    override fun handleMessage(event: MessageEvent, param: CustomParam): ServiceCallStatistic? {
        val fileName = "${param.uid}-${param.type}.png"
        val path = FILE_DIV_PATH.resolve(fileName)

        val imgBytes: ByteArray = if (param.url != null) try {
            webClient
                .get()
                .uri(param.url).retrieve()
                .bodyToMono(ByteArray::class.java).block()!!

        } catch (_: Exception) {
            throw IllegalStateException.Fetch("自定义图片")
        } else byteArrayOf()

        val profile = userProfileRepository.findTopById(param.uid) ?:
        UserProfileLite()
            .apply {
                this.userId = param.uid
                this.id = param.uid
            }

        if (param.url != null) {
            if (!Files.exists(FILE_DIV_PATH)) {
                try {
                    Files.createDirectory(FILE_DIV_PATH)
                } catch (e: IOException) {
                    log.error("自定义：创建目录失败", e)
                }
            }

            // 保存
            try {
                Files.write(path, imgBytes)
            } catch (e: Exception) {
                log.error("自定义：文件添加失败", e)
                event.reply("设置 ${param.type} 失败。错误已记录。")
                return ServiceCallStatistic.building(event)
            }

            profile.applyType(param.type, path.toAbsolutePath().toString())

            userProfileRepository.saveAndFlush(profile)
            event.reply("设置 ${param.type} 成功！")
            return ServiceCallStatistic.building(event)
        } else {
            // 删除
            try {
                Files.delete(path)
            } catch (_: NoSuchFileException) {
                event.reply("""
                    删除 ${param.type} 失败。
                    数据库里不存在你设置的自定义图片呢。
                """.trimIndent())
                return ServiceCallStatistic.building(event)
            } catch (e: Exception) {
                log.error("自定义：文件删除失败", e)
                event.reply("删除 ${param.type} 失败。错误已记录。")
                return ServiceCallStatistic.building(event)
            }

            profile.applyType(param.type, null)

            userProfileRepository.saveAndFlush(profile)
            event.reply("删除 ${param.type} 成功！")
            return ServiceCallStatistic.building(event)
        }
    }

    /**
     * 设置新的自定义配置, 并等待审核
     */
    fun pendingVerificationQueue(userId:Long, type: UserProfile.UserProfileType, path:Path) {
        val pathStr = path.toAbsolutePath().toString()
        val item = UserProfileItem(userId, type.column, pathStr)
        userProfileItemRepository.saveAndFlush(item)
    }

    /**
     * 删除某类型的自定义配置
     */
    fun removeProfileItem(userId: Long, type: UserProfile.UserProfileType) {
        userProfileItemRepository.deleteByUserIdAndType(UserProfileKey(userId, type.column))
    }

    /**
     * 删除所有的自定义配置
     */
    fun removeAllProfileItem(userId: Long) {
        userProfileItemRepository.deleteAllByUserId(userId)
    }

    internal enum class CustomOperate {
        ADD,
        DELETE,
        UNKNOWN,

        ;
        companion object {
            fun getOperate(input: String?): CustomOperate {
                return when (input) {
                    "s",
                    "save",
                    "a",
                    "add" -> ADD

                    "c",
                    "clear",
                    "d",
                    "delete",
                    "r",
                    "remove" -> DELETE

                    else -> UNKNOWN
                }
            }
        }
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(CustomService::class.java)
        lateinit var FILE_DIV_PATH: Path


        private fun UserProfileLite.applyType(type: UserProfile.UserProfileType, path: String? = null) {

            when (type) {
                CARD -> card = path
                MASCOT -> mascot = path
                AVATAR_FRAME -> avatarFrame = path

                INFO -> infoPanel = path
                SCORE -> scorePanel = path
                PPM -> ppmPanel = path

                else -> banner = path
            }
        }
    }
}
